File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists == null || lists.length == 0)
            return null;
        
        ListNode dummy = new ListNode(0);
        ListNode prev = dummy;
        
        PriorityQueue<ListNode> pq = new PriorityQueue<ListNode>(lists.length, new Comparator<ListNode>(){
            @Override
            public int compare(ListNode l1, ListNode l2){
                if(l1.val < l2.val)
                    return -1;
                else if(l1.val > l2.val)
                    return 1;
                else
                    return 0;
            }
        });
        
        for(ListNode node : lists){
            if(node != null)
                pq.add(node);
        }
        
        while(pq.size() > 0){
            prev.next = pq.poll();
            
            prev = prev.next;
            if(prev.next != null){
                pq.add(prev.next);
            }
        }
        
        return dummy.next;
    }
}

File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode MergeKLists(ListNode[] lists) {
        ListNode root = new ListNode(0);
        ListNode curr = root;
        PriorityQueue<ListNode,ListNode> pq = 
            new PriorityQueue<ListNode,ListNode>(Comparer<ListNode>.Create((x,y) => x.val.CompareTo(y.val)));
        
        foreach(ListNode list in lists){
            if(list != null)
                pq.Enqueue(list,list);
        }
        
        while(pq.Count > 0){
            ListNode node = pq.Dequeue();
            curr.next = new ListNode(node.val);
            curr = curr.next;
            
            node = node.next;
            if(node != null){
                pq.Enqueue(node,node);
            }
        }
        
        return root.next;
    }
}